src/ast.ml

```diff
@@ -15,99 +12,62 @@ let same_block_list_kind k1 k2 =
-    attributes: attributes;
+    term: 'a;
+    defs: 'a list;
   }
 
-module type T = sig
-  type t
-end
-
-module MakeBlock (I : T) = struct
-  type def_elt =
-    {
-      term: I.t;
-      defs: I.t list;
-    }
-
-  and block =
-    {
-      bl_desc: block_desc;
-      bl_attributes: attributes;
-    }
-
-  and block_desc =
-    | Paragraph of I.t
-    | List of list_type * list_spacing * block list list
-    | Blockquote of block list
-    | Thematic_break
-    | Heading of int * I.t
-    | Code_block of string * string
-    | Html_block of string
-    | Definition_list of def_elt list
-end
-
-type link =
+type ('a, 'attr) block =
+  | Paragraph of 'attr * 'a
+  | List of 'attr * list_type * list_spacing * ('a, 'attr) block list list
+  | Blockquote of 'attr * ('a, 'attr) block list
+  | Thematic_break of 'attr
+  | Heading of 'attr * int * 'a
+  | Code_block of 'attr * string * string
+  | Html_block of 'attr * string
+  | Definition_list of 'attr * 'a def_elt list
+
+type 'a link =
   {
-    label: inline;
+    label: 'a;
     destination: string;
     title: string option;
   }
 
-and inline =
-  {
-    il_desc: inline_desc;
-    il_attributes: attributes;
-  }
-
-and inline_desc =
-  | Concat of inline list
-  | Text of string
-  | Emph of inline
-  | Strong of inline
-  | Code of string
-  | Hard_break
-  | Soft_break
-  | Link of link
-  | Image of link
-  | Html of string
-
-module Raw = MakeBlock (String)
+type 'attr inline =
+  | Concat of 'attr * 'attr inline list
+  | Text of 'attr * string
+  | Emph of 'attr * 'attr inline
+  | Strong of 'attr * 'attr inline
+  | Code of 'attr * string
+  | Hard_break of 'attr
+  | Soft_break of 'attr
+  | Link of 'attr * 'attr inline link
+  | Image of 'attr * 'attr inline link
+  | Html of 'attr * string
 
-module Inline = struct type t = inline end
 
-include MakeBlock (Inline)
-
-module MakeMapper (Src : T) (Dst : T) = struct
-  module SrcBlock = MakeBlock(Src)
-  module DstBlock = MakeBlock(Dst)
-
-  let rec map (f : Src.t -> Dst.t) : SrcBlock.block -> DstBlock.block =
-    fun {bl_desc; bl_attributes} ->
-    let bl_desc =
-      match bl_desc with
-      | SrcBlock.Paragraph x -> DstBlock.Paragraph (f x)
-      | List (ty, sp, bl) ->
-          List (ty, sp, List.map (List.map (map f)) bl)
-      | Blockquote xs ->
-          Blockquote (List.map (map f) xs)
-      | Thematic_break ->
-          Thematic_break
-      | Heading (level, text) ->
-          Heading (level, f text)
-      | Definition_list l ->
-          let f {SrcBlock.term; defs} = {DstBlock.term = f term; defs = List.map f defs} in
-          Definition_list (List.map f l)
-      | Code_block (label, code) ->
-          Code_block (label, code)
-      | Html_block x ->
-          Html_block x
-    in
-    {bl_desc; bl_attributes}
-end
+type attributes =
+  (string * string) list
 
-module Mapper = MakeMapper (String) (Inline)
+type raw = (string, attributes) block
+type t = (attributes inline, attributes) block
```

Comment:
I can see the benefit of it. I don't feel strongly about it, but I would either
* keep it as is and wait until we have a need for functors (which is what I'd favour)
* do it in the previous style in which the constructors are part of a functor so that they are also namespaced